timers/timers.go

```diff
@@ -0,0 +1,255 @@
+// Package timers provides a timer facility similar to the standard library's.
+// It's designed to be faster, but less accurate.
+package timers
+
+import (
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/uber-go/atomic"
+)
+
+const cacheline = 64
+
+const (
+	// State machine for timers.
+	stateScheduled = iota
+	stateExpired
+	stateCanceled
+)
+
+// A Timer is a handle to a deferred operation.
+type Timer struct {
+	f        func()
+	deadline uint64 // in ticks
+	state    atomic.Int32
+	next     *Timer
+	tail     *Timer // only set on head
+}
+
+// Stop cancels the deferred operation. It returns whether or not the
+// cancellation succeeded.
+func (t *Timer) Stop() bool {
+	if t == nil {
+		return true // canceling a nil timer always succeeds
+	}
+	return t.state.CAS(stateScheduled, stateCanceled)
+}
+
+func (t *Timer) expire() bool {
+	return t.state.CAS(stateScheduled, stateExpired)
+}
+
+func (t *Timer) pushOne(head *Timer) *Timer {
+	head.next, head.tail = nil, nil
+	return t.push(head)
+}
+
+func (t *Timer) push(head *Timer) *Timer {
+	// link lists
+	if head == nil {
+		return t
+	}
+	if head.tail == nil {
```

Comment:
there's a lot of cases in this function, and i think a lot of them are to make `pushOne` work, so perhaps they should just be separate methods? something like:

```go
func (t *Timer) pushOne(node *Timer) *Timer {
  node.next = t
  if t == nil {
    node.tail = node
  } else {
    node.tail = t.tail
    t.tail = nil
  }
  return node
}

func (t *Timer) push(head *Timer) *Timer {
  if head == nil {
    return t
  }
  if t == nil {
    return head
  }

  head.next = t
  head.tail = t.tail
  t.tail = next
  return head
}
```

(I haven't tested the logic, but I think it's possible to write this with a lot less branching and complexity)